disnake/utils.py

```diff
@@ -1369,3 +1369,9 @@ def _overload_with_permissions(func: T) -> T:
 # this is used as a marker for functions or classes that were created by codemodding
 def _generated(func: T) -> T:
     return func
+
+
+# Similar to typing.assert_never, but returns instead of raising.
+# This is only to avoid "unreachable code", which pyright doesn't type-check.
+def assert_never(arg: Never, /) -> None:
```

Comment:
I'm not sure i fully understand how this is used and why it is under if typechecking blocks.

Reply:
It's a sanity check for these factory methods, to ensure that no component types are missed.

Take this for example:
https://github.com/DisnakeDev/disnake/blob/6f96e68481d49f96d97c37812620446465ef2d71/disnake/components.py#L482-L484
If everything is implemented correctly, `component_type` will be `Never` here, and the `assert_never` call hence type-checks fine.
If `data["type"]` were updated to `Literal[1, 2, 3, 4, 5]` (instead of `Literal[1, 2, 3, 4]`), `component_type` would be `Literal[5]` and the call would fail.

The type-checking branch is just there for performance reasons, since `assert_never` has no runtime effect (unlike the one in `typing`).

Hope that makes sense.

Reply:
I don't think the type-checking branch is too great, we might as well not check that we're typechecking unless you feel strongly about checking.

Reply:
Fair enough.
https://github.com/DisnakeDev/disnake/pull/801/commits/50ee58aea99ed7bb44d2edd68367a1f8f376276c